### PR TITLE
Add netlistTest function to Test.Tasty.Clash. Fixes #922

### LINKED ISSUE #922: Netlist Tests in clash-testsuite
Writing tests for #918 are better done by examining the AST of the netlist to be generated, rather than performing string matching on the generated HDL (as we care about the structure, and this is lost when HDL is written to a file).

I propose adding netlist to output tests, where the output is serialized Netlist instead of a particular HDL. It would then be possible to define a function (e.g. `mainNetlist`) for tests which care about looking at this (while still allowing other output tests as before).

### PR DESCRIPTION
**Pull request by alex-mckenna**
It is now possible to write test functions on the netlist AST
generated by clash-lib. The implementation here is slightly ugly,
and the majority has been moved to a separate module because of
this. The reason for this is that the clash executable does not
support stopping and outputting at an arbitrary stage (so clash-lib
is used directly instead).

Merging this PR will go towards allowing #918 to be adequately tested.

### CODE REVIEW

File: testsuite/Test/Tasty/Clash/NetlistTest.hs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+mkClashOpts :: [FilePath] -> ClashOpts
+mkClashOpts idirs = defClashOpts
+  { opt_cachehdl     = False
+  , opt_errorExtra   = True
+  , opt_floatSupport = True
+  , opt_importPaths  = idirs
+  }
+
+backend :: VHDLState
```
Comment (christiaanb) [inline testsuite/Test/Tasty/Clash/NetlistTest.hs]: Can we test this with the Verilog backend

Comment (alex-mckenna) [reply testsuite/Test/Tasty/Clash/NetlistTest.hs]: We can, does the choice of backend affect the generated netlist?

Comment (martijnbastiaan) [reply testsuite/Test/Tasty/Clash/NetlistTest.hs]: Yes, but usually minimally. Some blackboxes will have definitions that can be used inline in one language, but would require an assignment declaration in anther.

Comment (martijnbastiaan) [reply testsuite/Test/Tasty/Clash/NetlistTest.hs]: @christiaanb Did you mean "test with Verilog _too_", or would you like to discard the VHDL test altogether?

Comment (alex-mckenna) [reply testsuite/Test/Tasty/Clash/NetlistTest.hs]: Update: The current version allows running netlist tests against any backend, so I would say this is probably resolved.

Q. Can we test this with the Verilog backend?
A. Yes. Yes we can

File: testsuite/clash-testsuite.cabal [hunk 2]
```diff
@@ -1,62 +1,88 @@
...
+    -- it anyway so we're better off doing it beforehand.
+    clash-ghc,
+    clash-lib,
+    clash-prelude
+
   if flag(cosim)
     build-depends:     clash-cosim
 
-  -- hs-source-dirs:
-  default-language:    Haskell2010
```
Comment (martijnbastiaan) [inline testsuite/clash-testsuite.cabal]: Why is this removed?

Comment (alex-mckenna) [reply testsuite/clash-testsuite.cabal]: It isn't (but I admit the diff could be clearer). It's at the top of the `common` stanza

Comment (martijnbastiaan) [reply testsuite/clash-testsuite.cabal]: Woops, thanks.

**Comment by alex-mckenna** [comment]
Rebasing, so stressful. What do you think @martijnbastiaan?

**Comment by martijnbastiaan** [comment]
Looks great!

Don't fear rebasing, [`git reflog` has your back](http://gitready.com/intermediate/2009/02/09/reflog-your-safety-net.html)!
